builtin/credential/ldap/path_config.go

```diff
@@ -0,0 +1,143 @@
+package ldap
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathConfig(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: `config`,
+		Fields: map[string]*framework.FieldSchema{
+			"url": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "ldap URL to connect to (default: ldap://127.0.0.1)",
+			},
+			"domain": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "LDAP domain to use (eg: dc=example,dc=org)",
+			},
+			"userattr": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Attribute used for users (default: cn)",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:  b.pathConfigRead,
+			logical.WriteOperation: b.pathConfigWrite,
+		},
+
+		HelpSynopsis:    pathConfigHelpSyn,
+		HelpDescription: pathConfigHelpDesc,
+	}
+}
+
+func (b *backend) Config(req *logical.Request) (*ConfigEntry, error) {
+	entry, err := req.Storage.Get("config")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+	var result ConfigEntry
+	result.SetDefaults()
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (b *backend) pathConfigRead(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+
+	cfg, err := b.Config(req)
+	if err != nil {
+		return nil, err
+	}
+	if cfg == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"url":      cfg.Url,
+			"domain":   cfg.Domain,
+			"userattr": cfg.UserAttr,
+		},
+	}, nil
+}
+
+func (b *backend) pathConfigWrite(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+
+	cfg := &ConfigEntry{}
+	url := d.Get("url").(string)
+	if url != "" {
+		cfg.Url = strings.ToLower(url)
+	}
+	userattr := d.Get("userattr").(string)
+	if url != "" {
+		cfg.UserAttr = strings.ToLower(userattr)
+	}
+	domain := d.Get("domain").(string)
+	if url != "" {
+		cfg.Domain = domain
+	}
+
+	if !cfg.ValidateURL() {
+		return logical.ErrorResponse("LDAP URL is malformed"), nil
+	}
+
```

Comment:
It may be nice to try to connect to the LDAP endpoint after validation and just check that it is valid. This way the user gets an early fail-fast indicator.